other/n_body_simulation.py

```diff
@@ -0,0 +1,262 @@
+        gravitation_constant: float = 1.0,
+        time_factor: float = 1.0,
+        softening_factor: float = 0.0,
+    ) -> None:
+        self.bodies = bodies
+        self.gravitation_constant = gravitation_constant
+        self.time_factor = time_factor
+        self.softening_factor = softening_factor
+
+    def update_system(self: BodySystem, delta_time: float) -> None:
+        """
+        For each body, loop through all other bodies to calculate the total
+        force they exert on it. Use that force to update the body's velocity.
+
+        >>> body_system = BodySystem([Body(0,0,0,0), Body(10,0,0,0)])
+        >>> body_system.update_system(1)
+        >>> body_system.bodies[0].position_x
+        0.01
+        """
+        for body1 in self.bodies:
+            force_x = 0.0
+            force_y = 0.0
+            for body2 in self.bodies:
+                if body1 != body2:
+                    dif_x = body2.position_x - body1.position_x
+                    dif_y = body2.position_y - body1.position_y
+
+                    # Calculation of the distance using Pythagoras's theorem
+                    # Extra factor due to the softening technique
+                    distance = (dif_x ** 2 + dif_y ** 2 + self.softening_factor) ** (
+                        1 / 2
+                    )
+
+                    # Newton's law of universal gravitation.
+                    force_x += (
+                        self.gravitation_constant * body2.mass * dif_x / distance ** 3
+                    )
+                    force_y += (
+                        self.gravitation_constant * body2.mass * dif_y / distance ** 3
+                    )
+
+            # Update the body's velocity once all the force components have been added
+            body1.update_velocity(force_x, force_y, delta_time * self.time_factor)
+
+        # Update the positions only after all the velocities have been updated
+        for body in self.bodies:
+            body.update_position(delta_time * self.time_factor)
+
+
+def plot(
+    title: str,
+    body_system: BodySystem,
+    x_start: float = -1,
+    x_end: float = 1,
+    y_start: float = -1,
+    y_end: float = 1,
+) -> None:
+    """
+    Utility function to plot how the given body-system evolves over time.
+    No doctest provided since this function does not have a return value.
+    """
+
+    INTERVAL = 20  # Frame rate of the animation
+    DELTA_TIME = INTERVAL / 1000  # Time between time steps in seconds
+
+    fig = plt.figure()
+    fig.canvas.set_window_title(title)
+
+    # Set section to be plotted
+    ax = plt.axes(xlim=(x_start, x_end), ylim=(y_start, y_end))
+
+    # Each body is drawn as a patch by the plt-function
+    patches = []
+    for body in body_system.bodies:
+        patches.append(
+            plt.Circle((body.position_x, body.position_y), body.size, fc=body.color)
+        )
+
+    # Function called once at the start of the animation
+    def init() -> list[patches.Circle]:

```

Comment:
No doctest since this is an inner function.

Reply:
You mean that because it is an inner function it cannot contain bugs?

It needs tests.

Reply:
I'm not sure that it's possible to write a doctest for an inner function, for example, see https://stackoverflow.com/questions/2136910/can-i-unit-test-an-inner-function-in-python or https://bugs.python.org/issue1650090 . I could define the 2 functions outside the plot-function and add doctests. But that would be a little unintuitive, since they are just part of the plotting and not of the algorithm proper. But my experience with doctest is limited so I'm not sure what the best approach here is.

Reply:
I do not see the advantage to making it an inner function.  Just make it a normal function with proper tests.

Reply:
I ran into some problems while trying to define `init` & `update` outside the scope of `plot`, since these functions make use of `patches`, which is defined inside `plot`. Something like this seems to be the normal procedure for using matplotlib animation, for example, see `line` in https://jakevdp.github.io/blog/2012/08/18/matplotlib-animation-tutorial/ .

I don't think that we can pass `patches` as an argument to these functions since `animation.FuncAnimation` uses them as callbacks. It might be possible to solve this problem through currying, but that would make it needlessly complicated and may defy the attempt to avoid using inner functions. But maybe there is an easier solution that I'm missing.

Reply:
Another approach would be to define `patches` (& possibly other local variables needed) globally and then import them into the functions using the global-keyword. It should work but it's not pretty.

Reply:
If you want to keep the inner functions then the outer function should be heavily tested.  Let's put all calculation in a separate function from the plotting function.

Make the function that plots as small as possible... That is make one well tested function that calculates the results and another smaller, untested function that does no calculations but merely plots the results.

Reply:
That sounds like a good solution.